packages/provider/index.js

```diff
@@ -33,33 +34,37 @@ module.exports = {
   },
 
   testConnection: function(options) {
+    const { networks, network } = options;
+    let networkCheckTimeout;
+    if (networks[network]) {
+      networkCheckTimeout =
+        networks[network].networkCheckTimeout || DEFAULT_NETWORK_CHECK_TIMEOUT;
+    } else {
+      networkCheckTimeout = DEFAULT_NETWORK_CHECK_TIMEOUT;
+    }
     const provider = this.getProvider(options);
```

Comment:
still think calling `getProvider` and instantiating a new `InterfaceAdapter/Web3Shim` here is redundant, esp considering the rest of `Environment.detect` uses the already instantiated adapter/shim in the [helpers](https://github.com/trufflesuite/truffle/blob/develop/packages/environment/environment.js#L25-L26).

Reply:
One thing that using `getProvider` gives us is more flexibility with what `options` can be. Not using it to get the provider makes it so that the `provider` property will need to be a provider and it will not accept a `host` and `port` property etc. I'm not sure what behavior we want for this method. I'm not sure I have a strong preference, any thoughts?

Reply:
The getter for `config.provider` will already return [Provider.create](https://github.com/trufflesuite/truffle/pull/2510/files/58c594d5693f2927fa4769758df927672dc5f749#diff-eb621b6113cc100218116be01feae659L11) which calls `getProvider` & returns a normalized Provider. This also happens right before the `testConnection` call in `Environment.detect`.

It makes sense that `testConnection` _should_ be testing an already normalized Provider.
Also makes sense to DRY up the code in general.

Reply:
Yes, I understand what you are saying that it should have already parsed and set a provider on the config object. However what I am pointing out is that `testConnection` is exposed in this package. The `getProvider` method allows for the input to be a little more flexible.

I don't know what you mean by DRY here because bye using the `getProvider` method you are re-using existing code.

Reply:
Ah, I see. I agree as an exposed pkg util func it's useful. It still feels repetitive to implement it in the `Environment.detect` flow.